common/lib/dependabot/clients/bitbucket_server.rb

```diff
@@ -0,0 +1,236 @@
+        url = File.join(repo_path, "/branches")
+        body = {
+          name: branch_name,
+          startPoint: base_commit
+        }
+
+        post(url, body)
+      end
+
+      # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def fetch_repo_contents(_repo_name, commit = nil, path = nil)
+        raise "Commit is required if path provided!" if commit.nil? && path
+
+        api_path = "browse"
+        api_path += "/#{path.gsub(%r{/+$}, '')}" if path
+        api_path += "?at=#{commit}&" if commit
+        api_path += path || commit ? "limit=100" : "?limit=100"
+        response = get(File.join(repo_path, api_path))
+
+        response.body.dig("children", "values")&.map do |file|
+          file["path"] = if path.nil? || path.empty?
+                           file.dig("path", "name")
+                         else
+                           path
+                         end
+          file
+        end
+      end
+
+      # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def fetch_file_contents(_repo_name, commit, path)
+        path = "raw/#{path.gsub(%r{/+$}, '')}?at=#{commit}"
+        response = get(File.join(repo_path, path))
+        response.body
+      end
+
+      def tags(_repo_name)
+        url = File.join(repo_path, "tags?limit=100")
+        response = get(url)
+
+        response.body.fetch("values")
+      end
+
+      def compare(_repo_name, previous_tag, new_tag)
+        path = "compare/commits?to=#{new_tag}&from=#{previous_tag}"
+        response = get(File.join(repo_path, path))
+
+        response.body.fetch("values")
+      end
+
+      def fetch_pull_request_for_branch(branch)
+        path = "/pull-requests?at=refs/heads/#{branch}&direction=outgoing"
+        url = File.join(repo_path, path)
+        response = get(url)
+
+        response.body.dig("values", 0)
+      end
+
+      def create_pull_request(title, from, to, description, reviewers = nil)
+        url = File.join(repo_path, "pull-requests")
+        from_ref = { "id" => from }
+        to_ref   = { "id" => to }
+        body = {
+          title: title,
+          description: description,
+          state: "OPEN",
+          fromRef: from_ref,
+          toRef: to_ref,
+          reviewers: Array(reviewers).map { |r| { user: { name: r } } }
+        }.to_json
+        post(url, body)
+      end
+
+      private
+
+      attr_reader :credentials
+      attr_reader :source
+
+      def request(method, url, body = nil)
+        res = conn.public_send(method) do |req|
+          req.url url
+          req.headers["Content-Type"] = "application/json"
+          req.body = body unless body.nil?
+        end
+
+        handle_response(res)
+      end
+
+      def handle_response(response)
+        raise Unauthorized if response.status == 401
+        raise Forbidden if response.status == 403
+        raise NotFound if response.status == 404
+
+        if response.status >= 400
```

Comment:
I've received a 500 internal server error for a specific pull request bumping simplecov 0.16.1 -> 0.18.5.

Somewhat hilariously, the changelog diff contains the :see_no_evil: (`:see_no_evil:`) emoji, which my Bitbucket server won't let me insert into the database:

```
Exception executing batch [java.sql.BatchUpdateException: Incorrect string value: '\xF0\x9F\x99\x88 F...' for column 'description' at row 1], SQL: insert into sta_pull_request (closed_timestamp, created_timestamp, description, from_branch_name, from_hash, from_branch_fqn, from_repository_id, locked_timestamp, rescoped_timestamp, scoped_id, pr_state, title, to_branch_name, to_hash, to_branch_fqn, to_repository_id, updated_timestamp, entity_version, id) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
```

`\xF0\x9F\x99\x88` is the four byte sequence of 1U+1F648`, which is this emoji.

This might be because this server is quite old and has been upgraded multiple times. It's currently on 5.16.4, using MySQL 5.6, and using the `utf8` character set for the table, not `utf8_mb4`. It looks from [this](https://jira.atlassian.com/browse/BSERV-8490) that even the latest Bitbucket Server won't support higher.

Perhaps the best way to improve this without changing the other clients would be a kind of hook system - where clients can be notified on various events to mutate the date (like stripping four byte unicode chars out of the PR, in this case). The other issue I've been having is that dependabot can produce git refs over 100 characters produced by the branch namer, which a hook could similarly be used for?